File: solutions/java/collatz-conjecture/1/src/main/java/CollatzCalculator.java
public final class CollatzCalculator {
    public int computeStepCount(int start) {
        if (start < 1)
            throw new IllegalArgumentException("Only positive integers are allowed");
        
        return computeStepCount(start, 0);
    }

    private int computeStepCount(int start, int steps) {
        if (start == 1)
            return steps;

        int next = start % 2 == 0
            ? start / 2 : start * 3 + 1;
        
        return computeStepCount(next, steps + 1);
    }
}

File: solutions/kotlin/collatz-conjecture/1/src/main/kotlin/CollatzCalculator.kt
object CollatzCalculator {
    fun computeStepCount(start: Int): Int {
        require(start > 0) {
            "Input number must be positive."
        }

        return computeStepCount(start, 0)
    }

    private tailrec fun computeStepCount(n: Int, count: Int): Int = 
        when {
            n == 1 -> count
            n.isEven() -> computeStepCount(n / 2, count + 1)
            else -> computeStepCount(n * 3 + 1, count + 1)
        }

    private fun Int.isEven() = this % 2 == 0
}
